--- pth.h.in.dist	2002-01-27 12:03:40.000000000 +0100
+++ pth.h.in	2002-07-23 03:20:19.000000000 +0200
@@ -50,6 +50,14 @@
 @FALLBACK_PID_T@
 @FALLBACK_SIZE_T@
 @FALLBACK_SSIZE_T@
+/*
+ * CYGWIN pulls a trick on the preprocessor for the socklen_t type.
+ * Until AC_CHECK_TYPEDEF macro in aclocal.m4 is aware of this, this
+ * is a work-around to disable the warning
+ */
+#ifdef __CYGWIN__
+#undef socklen_t
+#endif
 @FALLBACK_SOCKLEN_T@
 @FALLBACK_OFF_T@
 @FALLBACK_SIG_ATOMIC_T@
